protocols/identify/src/behaviour.rs

```diff
@@ -40,6 +41,50 @@ use std::{
     time::Duration,
 };
 
+/// Whether an [`Multiaddr`] is a valid for the QUIC transport.
+fn is_quic_addr(addr: &Multiaddr, v1: bool) -> bool {
+    use Protocol::*;
+    let mut iter = addr.iter();
+    let Some(first) = iter.next() else {
+        return false;
+    };
+    let Some(second) = iter.next() else {
+        return false;
+    };
+    let Some(third) = iter.next() else {
+        return false;
+    };
+    let fourth = iter.next();
+    let fifth = iter.next();
+
+    matches!(first, Ip4(_) | Ip6(_) | Dns(_) | Dns4(_) | Dns6(_))
+        && matches!(second, Udp(_))
+        && if v1 {
+            matches!(third, QuicV1)
+        } else {
+            matches!(third, Quic)
+        }
+        && matches!(fourth, Some(P2p(_)) | None)
+        && fifth.is_none()
+}
```

Comment:
I think this came up another time already. Why do we need to perform address translation for QUIC addresses? QUIC never uses ephemeral ports (to my knowledge).

Reply:
Autonat server should use ephemeral port when dialing back a peer.
otherwise the server dial-back will always succeed (and will report a false positive to the peer) because the peer has already opened the states in the firewalls when dialing the server.
=> translation for QUIC addresses is required

@umgefahren : When dialing back,  the server should use a **dedicated** quic endpoint (not localy binded) and should not reuse a listner endpoint (this is now not the case)

Reply:
> Autonat server should use ephemeral port when dialing back a peer. otherwise the server dial-back will always succeed (and will report a false positive to the peer) because the peer has already opened the states in the firewalls when dialing the server. => translation for QUIC addresses is required
> 
> @umgefahren : When dialing back, the server should use a **dedicated** quic endpoint (not localy binded) and should not reuse a listner endpoint (this is now not the case)

You are right. Thank you for laying it out for me. We need to do address translation because **we** (libp2p) decide to **explicitly** use an ephemeral port for a new QUIC connection to avoid accidental hole-punching despite QUIC itself not needing it.

I think this is worthwhile documenting here because anybody coming across this will think "hey, QUIC always uses stable ports so why would be bother with address translation for QUIC". (Also cc @mxinden, I know you too have been confused by this in the past)

Reply:
> Autonat server should use ephemeral port when dialing back a peer. otherwise the server dial-back will always succeed (and will report a false positive to the peer) because the peer has already opened the states in the firewalls when dialing the server. => translation for QUIC addresses is required
> 
> @umgefahren : When dialing back, the server should use a **dedicated** quic endpoint (not localy binded) and should not reuse a listner endpoint (this is now not the case)

I will change it so that when Endpoint is Dialer and PortUse::Reuse we dial a dedicated quic endpoint.